cmd/rocsp-tool/client.go

```diff
@@ -103,26 +118,35 @@ func (cl *client) loadFromDB(ctx context.Context, speed ProcessingSpeed) error {
 // its output channel at a maximum frequency of `frequency`. When it's read all available rows, it
 // closes its output channel and exits.
 // If there is an error, it logs the error, closes its output channel, and exits.
-func (cl *client) scanFromDB(ctx context.Context, minID int64, frequency time.Duration, inflightIDs *inflight) <-chan *sa.CertStatusMetadata {
+func (cl *client) scanFromDB(ctx context.Context, minID int64, maxID int64, frequency time.Duration, inflightIDs *inflight) <-chan *sa.CertStatusMetadata {
 	statusesToSign := make(chan *sa.CertStatusMetadata)
 	go func() {
 		defer close(statusesToSign)
-		err := cl.scanFromDBInner(ctx, minID, frequency, statusesToSign, inflightIDs)
-		if err != nil {
-			log.Printf("error scanning rows: %s", err)
+
+		var err error
+		currentMin := minID
+		for currentMin < maxID {
+			currentMin, err = cl.scanFromDBOneBatch(ctx, minID, frequency, statusesToSign, inflightIDs)
+			if err != nil {
+				log.Printf("error scanning rows: %s", err)
+			}
 		}
 	}()
 	return statusesToSign
 }
 
-func (cl *client) scanFromDBInner(ctx context.Context, minID int64, frequency time.Duration, output chan<- *sa.CertStatusMetadata, inflightIDs *inflight) error {
+// scanFromDBOneBatch scans up to `cl.scanBatchSize` rows from certificateStatus, in order, and
+// writes them to `output`. When done, it returns the highest `id` it saw during the scan.
+// We do this in batches because if we tried to scan the whole table in a single query, MariaDB
+// would terminate the query after a certain amount of data transferred.
+func (cl *client) scanFromDBOneBatch(ctx context.Context, minID int64, frequency time.Duration, output chan<- *sa.CertStatusMetadata, inflightIDs *inflight) (int64, error) {
 	rowTicker := time.NewTicker(frequency)
 
-	query := fmt.Sprintf("SELECT %s FROM certificateStatus WHERE id >= ?",
+	query := fmt.Sprintf("SELECT %s FROM certificateStatus WHERE id > ? ORDER BY id LIMIT ?",
```

Comment:
Changing this to `>` makes sense for subsequent batches (because the minID being passed in is the maxID we've already received, and therefore don't need to retrieve again). But it also means that the very first batch will exclude its lower limit, which is rather confusing/counterintuitive behavior, especially for software engineers who are used to half-open intervals facing the other direction.

Reply:
yeah, same. Proposal here is to do inclusive for the first batch and exclusive for subsequent?

Reply:
I guess? I don't love that either, since it makes the code more complex. I think maybe the better proposal is to change the name of `minID` to `prevID` or something like that, so that folks reading the code have a different semantic expectation in mind when they get to this query?

Reply:
Discussed offline; going to merge without addressing this nit.